### ISSUE TITLE
Could not read from remote repository.

### ISSUE THREAD

**Issue by srheaume** [concourse/concourse#6544]
## Summary

I upgraded from version 6.5.5 to version 7.0.0.

Since then, I am unable to clone a git resource. I use a private SSH key for repo authorization.

```
resources:
- icon: gitlab
  name: concourse-ci-pipelines
  source:
    branch: develop
    private_key: (("gitlab.xyz.com.ssh_private_key))
    uri: git@gitlab.xyz.com:summit/common/tools/concourse-ci-pipelines.git
```

I redeployed Concourse v6.5.5 with the same deployment scripts. It works very well. I start the procedure again by changing only the version number to v7.0.0 and I still get this result:

> selected worker: prd-concourse-worker-67449cbb77-ct77j
> Identity added: /tmp/git-resource-private-key
> Cloning into '/tmp/git-resource-repo-cache'...
> fatal: Could not read from remote repository.
> 
> Please make sure you have the correct access rights
> and the repository exists.

Do you have an idea how to diagnose this problem?

**Comment by marco-m** [comment]
I see a missing closing quote `"`:

```
private_key: (("gitlab.xyz.com.ssh_private_key))
```

A way to pinpoint when the change happened is to compare the version of the git resource that is shipped with 6.5.5 with the version of the git resource that is shipped with 7.0.0.

if you log into a Concourse worker you should see in the home directory of the worker a series of directories; in one of them (if I am remembering correctly) there should be the embedded resources, among them the git one. From there, you can see the version.

**Comment by srheaume** [comment]
# More Details

Just to clarify, all our environments are IaC. So, Concourse and the cluster it resides in gets rebuilt at each update. So, there are no possible leftovers from the old version. The problem seems to be related to the Git repository access through SSH.

# Installation

The installation of Concourse is done using the helm chart version v14.6.0 which corresponds to the version v7.0.0 of the application.

# Pipeline Code

``` yaml
jobs:
- name: run-garbage-collection
  plan:
  - in_parallel:
      steps:
      - get: concourse-ci-pipelines
        trigger: true
      - get: summit-common-manifests-ci
        trigger: true
      - get: summit-odience-manifests-ci
        trigger: true
  - config:
      image_resource:
        name: ""
        source:
          repository: summittech/concourse-ci-tools
        type: registry-image
      inputs:
      - name: concourse-ci-pipelines
      - name: summit-common-manifests-ci
      - name: summit-odience-manifests-ci
      params:
        CONCOURSE_PASSWORD: ((concourse-ci-pipelines.password))
        CONCOURSE_URL: ((concourse-ci-pipelines.url))
        CONCOURSE_USERNAME: ((concourse-ci-pipelines.username))
      platform: linux
      run:
        args:
        - summit-common-manifests-ci
        - summit-odience-manifests-ci
        dir: concourse-ci-pipelines
        path: ./pipelines/master-factory/task-scripts/collect-project-pipelines.sh
    task: collect-project-pipelines
  - config:
      image_resource:
        name: ""
        source:
          repository: summittech/concourse-ci-tools
        type: registry-image
      inputs:
      - name: concourse-ci-pipelines
      params:
        CONCOURSE_PASSWORD: ((concourse-ci-pipelines.password))
        CONCOURSE_URL: ((concourse-ci-pipelines.url))
        CONCOURSE_USERNAME: ((concourse-ci-pipelines.username))
      platform: linux
      run:
        dir: concourse-ci-pipelines
        path: ./pipelines/master-factory/task-scripts/collect-teams.sh
    task: collect-teams
- name: set-project-factories.summit-common-manifests-ci
  plan:
  - in_parallel:
      steps:
      - get: concourse-ci-pipelines
        trigger: true
      - get: summit-common-manifests-ci
        trigger: true
  - config:
      image_resource:
        name: ""
        source:
          repository: summittech/concourse-ci-tools
        type: registry-image
      inputs:
      - name: concourse-ci-pipelines
      - name: summit-common-manifests-ci
      params:
        CONCOURSE_PASSWORD: ((concourse-ci-pipelines.password))
        CONCOURSE_URL: ((concourse-ci-pipelines.url))
        CONCOURSE_USERNAME: ((concourse-ci-pipelines.username))
      platform: linux
      run:
        args:
        - summit-common-manifests-ci
        dir: concourse-ci-pipelines
        path: ./pipelines/master-factory/task-scripts/set-project-factories.sh
    task: set-project-factories
- name: set-project-factories.summit-odience-manifests-ci
  plan:
  - in_parallel:
      steps:
      - get: concourse-ci-pipelines
        trigger: true
      - get: summit-odience-manifests-ci
        trigger: true
  - config:
      image_resource:
        name: ""
        source:
          repository: summittech/concourse-ci-tools
        type: registry-image
      inputs:
      - name: concourse-ci-pipelines
      - name: summit-odience-manifests-ci
      params:
        CONCOURSE_PASSWORD: ((concourse-ci-pipelines.password))
        CONCOURSE_URL: ((concourse-ci-pipelines.url))
        CONCOURSE_USERNAME: ((concourse-ci-pipelines.username))
      platform: linux
      run:
        args:
        - summit-odience-manifests-ci
        dir: concourse-ci-pipelines
        path: ./pipelines/master-factory/task-scripts/set-project-factories.sh
    task: set-project-factories
resources:
- icon: gitlab
  name: concourse-ci-pipelines
  source:
    branch: develop
    password: (("gitlab03.summit-tech.org".password))
    private_key: (("gitlab03.summit-tech.org".ssh_private_key))
    uri: git@gitlab03.summit-tech.org:summit/common/tools/concourse-ci-pipelines.git
    username: (("gitlab03.summit-tech.org".username))
  type: git
- icon: clipboard-text
  name: summit-common-manifests-ci
  source:
    branch: develop
    password: (("gitlab03.summit-tech.org".password))
    private_key: (("gitlab03.summit-tech.org".ssh_private_key))
    uri: git@gitlab03.summit-tech.org:summit/common/manifests/ci.git
    username: (("gitlab03.summit-tech.org".username))
  type: git
- icon: clipboard-text
  name: summit-odience-manifests-ci
  source:
    branch: develop
    password: (("gitlab03.summit-tech.org".password))
    private_key: (("gitlab03.summit-tech.org".ssh_private_key))
    uri: git@gitlab03.summit-tech.org:summit/odience/manifests/ci.git
    username: (("gitlab03.summit-tech.org".username))
  type: git
```

**Comment by Roadmaster** [comment]
I'm seeing the same thing after a 7.0.0 upgrade. My pipelines are very simple, here's the relevant resource declaration:
```
resources:
  - name: resource-lalala
    type: git
    source:
      uri: git+ssh://my-user@git.launchpad.net/~user/project/+git/repo
      branch: master
      private_key: ((my-private-key))
      git_config:
          - name: core.sshCommand
            value: "ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no"
```

Happy to provide more info and/or dig further, if you let me know where to dig ;)

- Daniel

**Comment by taylorsilva** [comment]
I tried the following and was unable to reproduce this issue using `concourse/concourse:7.0.0` image.

```yaml
resources:
- name: concourse-docs-git
  type: git
  icon: github
  source:
    uri: git@github.com:concourse/docs.git
    # tried with and without specifying the branch
    branch: master
    # tried with and without git_config
    git_config:
          - name: core.sshCommand
            value: "ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no"
    private_key: |
      -----BEGIN OPENSSH PRIVATE KEY-----
      ......
      -----END OPENSSH PRIVATE KEY-----
    # these are not used since the uri is ssh, not http/https
    # tried with and without, still worked
    username: ignoredUser
    password: ignoredPassword


jobs:
- name: job
  public: true
  plan:
  - get: concourse-docs-git
    trigger: true
  - task: list-files
    config:
      inputs:
        - name: concourse-docs-git
      platform: linux
      image_resource:
        type: registry-image
        source: { repository: busybox }
      run:
        path: ls
        args: ["-la", "./concourse-docs-git"]
```

Not sure what I'm missing here. Is anyone able to make this(^) example break on their concourse? @srheaume @Roadmaster 

@srheaume I noticed that in your pipeline, you specify `username` and `password` for your git resources when you're using an ssh uri. Only `private_key` is needed in this case. This shouldn't matter but it's extra info in your pipeline that you don't need.

**Comment by srheaume** [comment]
@taylorsilva I specify the username and password in case I have Git repos accessible with HTTPS. The values that are in Vault are empty strings. As you said, it does not matter. I still took the time to try without these values; same result. You tell us that it works for you. Did you start with a fresh installation or an upgrade?

**Comment by taylorsilva** [comment]
I only tried locally on my machine using docker-compose with this [docker-compose file](https://concourse-ci.org/docker-compose.yml). I made sure to `docker pull concourse/concourse:7.0.0` before starting.

In our [Concourse instance](https://ci.concourse-ci.org/) though, we use ssh + private keys also in vault: https://github.com/concourse/ci/blob/8868b05b124e1b2389e2b06bcd0b8ef2a73c2bc9/pipelines/concourse.yml#L1580-L1586
and this is working fine for us. Therefore, I'm not convinced that vault integration is the issue.

#6115 is the only change to vault in 7.0.0

If you take the pipeline yaml I posted, and run it on your Concourse, does it work or do you get the same error?

Another thing you can try, use an older version of the git resource in one of your pipelines where it's broken.
Add the following to your pipeline. Concourse 7 shipped with git-resource 1.12.0, so the below uses 1.11.0.
```yaml
resource_types:
- name: git
  type: registry-image
  source:
    repository: concourse/git-resource
    tag: 1.11.0
```

You can keep going back until you get to a version that works for you: https://github.com/concourse/git-resource/releases
Let us know when it breaks/works, that'll help us track the problem down since I currently don't know what the issue could be.

**Comment by srheaume** [comment]
@taylorsilva Thank you for your reply. I think we're starting to narrow down the problem. It works fine with the following:

**Comment by taylorsilva** [comment]
Okay, just to sanity check, can you change it to 1.12.0, but still using the redefined resource type:
```yaml
resource_types:
- name: git
  type: registry-image
  source:
    repository: concourse/git-resource
    tag: 1.12.0
```

I still can't reproduce it myself and release notes for 1.12.0 don't seem at all related to the problem

> Allow for returning more than just the latest version in a check

**Comment by srheaume** [comment]
With the tag set 1.12.0, it does not work. Maybe is related to the following:

```
SSH_ASKPASS_REQUIRE=force
```

This has been introduced in version 1.12.0

**Comment by Roadmaster** [comment]
Um my broken pipelines are now magically working again :thinking: I did nothing except perhaps restart the worker (I checked right after restarting, broken, went to lunch, and when I came back they were working fine, fetching git repos like nothing).

**Comment by srheaume** [comment]
This does not work for me:

```yaml
---
resource_types:
- name: git
  type: registry-image
  source:
    repository: concourse/git-resource
    tag: 1.12.0

resources:
- name: summit-common-helm_chart-microservice-git
  type: git
  icon: github
  source:
    uri: git@github.com:srheaume/summit-common-helm_chart-microservice.git
    branch: develop
    private_key: (("github.com".ssh_private_key))

jobs:
- name: job
  public: true
  plan:
  - get: summit-common-helm_chart-microservice-git
    trigger: true
  - task: list-files
    config:
      inputs:
        - name: concourse-docs-git
      platform: linux
      image_resource:
        type: registry-image
        source: { repository: busybox }
      run:
        path: ls
        args: ["-la", "./concourse-docs-git"]
```

If I change the tag for 1.11.0, it works. I can't be simpler.

**Comment by srheaume** [comment]
> Um my broken pipelines are now magically working again  I did nothing except perhaps restart the worker (I checked right after restarting, broken, went to lunch, and when I came back they were working fine, fetching git repos like nothing).

Maybe I should go out for lunch more often ;-)

**Comment by taylorsilva** [comment]
I don't think `SSH_ASKPASS_REQUIRE=force` is the problem, that just results in a script getting called if the private key is password protected, which is not the case here.

Could you try hijacking into the check container (for the 1.12.0 version) and try cloning the repo manually, similar to what the `check.sh` script does. Like this

```sh
# url is the page of the resource where you can trigger a new check
$ fly -t dev i -u http://localhost:8080/teams/main/pipelines/test/resources/concourse-docs-git
bash-5.1# eval $(ssh-agent)
bash-5.1# ssh-add /tmp/git-resource-private-key
Identity added: /tmp/git-resource-private-key
bash-5.1# git clone git@github.com:srheaume/summit-common-helm_chart-microservice.git
```

This should help tell us if it's an issue with the scripting or not... kinda...

**Comment by srheaume** [comment]
All right ... Here are the results:

1) with tag set to 1.11.0 OK
2) with tag set to 1.12.0 FAILED

**Comment by srheaume** [comment]
Got it!!! This is a DNS issue. github.com can't be resolved.

resolv.conf (version 1.11.0):
```
nameserver 172.17.0.12

```

resolv.conf (version 1.12.0):
```
nameserver 172.17.0.12

```

The resolv.conf dile for both versions has the same content.

**Comment by Roadmaster** [comment]
Maybe try this:

I followed the recommendation to enable containerd:
`
CONCOURSE_RUNTIME=containerd
`

then I remembered I need to force the DNS, which was previously done in garden-config.ini, but the release notes suggest doing this now:
`
CONCOURSE_CONTAINERD_DNS_SERVER=192.168.1.1
`
so I think my sequence was:

 1. set CONCOURSE_RUNTIME=containerd
 2. restart, things don't work
 3. set the CONCOURSE_CONTAINERD_DNS_SERVER to what I had previously in garden-config.ini
 4. Somehow bork the worker restart in a way that made me think things weren't yet working
 5.  go to lunch
 6. worker restart completed properly, things are now working because of the DNS setting
 
for clarity, the reason I'm setting the DNS server this way is because of what's mentioned in https://concourse-ci.org/concourse-worker.html#troubleshooting-and-fixing-dns-resolution (which might need to be updated for containerd now?)

**Comment by taylorsilva** [comment]
![dns haiku](https://user-images.githubusercontent.com/4583837/108270429-6f595000-713d-11eb-9d4e-b37d9ab414e1.png)

**Comment by srheaume** [comment]
I will try to find out how to do the same using the Concourse Helm Chart.

**Comment by srheaume** [comment]
I can't find a config that works for me. The Helm Chart documentation does not help as well.

**Comment by srheaume** [comment]
@taylorsilva I set the runtime value to containerd and specify the value of my server and it works partially. I'll explain why.

Now, I have the Git resource that is able to clone repos without any problem. On the other hand, I have pipelines that make calls to the fly command whose Concourse URL is a Kubernets address (e.g. "http://concourse-web.concourse.svc.cluster.local:8080"). So those pipelines don't work anymore.

In summary, the new version of Concourse is hardly usable in a Kubernetes context.

**Comment by srheaume** [comment]
@taylorsilva One way to reproduce the problem is to do the deployment in Kubernetes. It can be minikube. Docker-composes, will never reproduce the networking environment of Kubernetes. After a discussion with a group of DevOps in Montreal. We are finally several of us with the same problem. What we have in common is that we use Kubernetes to deploy Concourse. Is there a plan to fix it? Should we open a new bug with the real description of the issue?

**Comment by taylorsilva** [comment]
@srheaume also yes, if you can open a new issue that better describes the problem with reproducible steps that would be awesome 🙏

**Comment by srheaume** [comment]
> Thanks for all the new info! That's really helpful. I'll try and reproduce with minikube and/or one of our GKE clusters. I have some questions that if you can answer will help me:
> 
> * k8 version this is happening on?
> * To clarify:
>   
>   * this issue started with the garden/guardian runtime? And...
The issue started with release 7.0.0 with default runtime.
>   * containerd kinda fixes the issue, the caveat being that containerd isn't enabling local k8s dns addresses to resolve?
Exactly
> * Any special config you had to do for the helm chart or did default values reproduce the issue?
Default values will reproduce the issue

**Comment by taylorsilva** [comment]
I'm going to close this issue. Thanks for all the details in the new issue you opened @srheaume! Should be easy for us to reproduce. Not sure when we'll fix it but I'll bring it up to the team.

**Comment by brh28** [comment]
Seems I'm facing the same issue described in this thread. We have concourse deployed to Kubernetes with version 7.11.2 and the Containerd runtime. 

@taylorsilva Is there a status update on a fix?
